src/lib/io/sol-memmap-storage.c

```diff
@@ -343,19 +355,123 @@ check_map(const struct sol_memmap_map *map)
     return true;
 }
 
+#ifdef USE_I2C
+static char *
+sol_slice_to_str(const struct sol_str_slice *slice)
+{
+    char *result = calloc(1, slice->len + 1);
+
+    SOL_NULL_CHECK(result, NULL);
+
+    memcpy(result, slice->data, slice->len);
+
+    return result;
+}
+#endif
+
+static int
+resolve_i2c_path(struct sol_memmap_map *map)
+{
+#ifndef USE_I2C
+    SOL_WRN("No support for i2c");
+    return -ENOTSUP;
+#else
+    char *rel_path = NULL, *dev_number_s = NULL, *dev_name = NULL, *end_ptr;
+    unsigned int dev_number;
+    struct sol_vector instructions;
+    struct sol_buffer result_path = SOL_BUFFER_INIT_EMPTY;
+    struct sol_str_slice command = sol_str_slice_from_str(map->path);
+    int ret = -EINVAL;
+
+    instructions = sol_util_str_split(command, ",", 5);
+    if (instructions.len < 5) {
+        SOL_WRN("Invalid create device path. Expected 'create,i2c,<rel_path>,"
+            "<devnumber>,<devname>'");
+        goto end;
+    }
+
+    rel_path = sol_slice_to_str(sol_vector_get(&instructions, REL_PATH_IDX));
+    SOL_NULL_CHECK_GOTO(rel_path, end);
+
+    dev_number_s = sol_slice_to_str(sol_vector_get(&instructions, DEV_NUMBER_IDX));
+    SOL_NULL_CHECK_GOTO(dev_number_s, end);
+
+    errno = 0;
+    dev_number = strtoul(dev_number_s, &end_ptr, 0);
+    if (errno || *end_ptr != '\0')
+        goto end;
+
+    dev_name = sol_slice_to_str(sol_vector_get(&instructions, DEV_NAME_IDX));
+    SOL_NULL_CHECK_GOTO(dev_name, end);
+
+    ret = sol_i2c_create_device(rel_path, dev_name, dev_number,
+        &result_path);
+
+    if (ret >= 0 || ret == -EEXIST) {
+        const struct sol_str_slice ending = SOL_STR_SLICE_LITERAL("/eeprom");
+        struct timespec start;
+        struct stat st;
+
+        /* Giving some time to sysfs update. Maybe listen for udev events? */
+        ret = sol_buffer_append_slice(&result_path, ending);
+        if (ret < 0)
+            goto end;
+
+        map->resolved_path = sol_buffer_steal(&result_path, NULL);
+
+        ret = 0;
+        start = sol_util_timespec_get_current();
+        while (stat(map->resolved_path, &st)) {
+            struct timespec elapsed, now = sol_util_timespec_get_current();
+
+            sol_util_timespec_sub(&now, &start, &elapsed);
+            /* Let's wait up to one second */
+            if (elapsed.tv_sec > 0) {
+                ret = -ENODEV;
+                goto end;
+            }
+        }
+    }
+
+end:
+    free(rel_path);
+    free(dev_number_s);
+    free(dev_name);
+    sol_vector_clear(&instructions);
+
+    return ret;
+#endif
+}
+
 SOL_API int
-sol_memmap_add_map(const struct sol_memmap_map *map)
+sol_memmap_add_map(struct sol_memmap_map *map)
 {
+    SOL_NULL_CHECK(map, -EINVAL);
+
+    if (strstartswith(map->path, "create,i2c,")) {
```

Comment:
But `entries` is actually an array at the end of the struct, not a pointer to an array (so it's possible to use `sol_str_table`). So I guess it'd be necessary to duplicate the entries, or use something as a dictionary.